author: Andres Salomon <dilinger@debian.org>
description: work around FTBFS on clang-13 under armhf & i386

We see the following on 32bit platforms on older clang:

ld.lld-13: error: undefined symbol: __mulodi4
>>> referenced by const_eval.cc


This symbol is in libclang_rt's builtins, but for whatever reason it doesn't
get included. Not worth sending upstream since this doesn't happen in
bookworm, only in bullseye.

--- a/third_party/dawn/src/tint/BUILD.gn
+++ b/third_party/dawn/src/tint/BUILD.gn
@@ -91,6 +91,15 @@ config("tint_public_config") {
     defines += [ "TINT_BUILD_IR=0" ]
   }
 
+  # work around a bug with clang not being able to find the __mulodi4 symbol on 32-bit archs
+  if (current_cpu == "x86") {
+    lib_dirs = [ "${sysroot}/usr/lib/llvm-13/lib/clang/13.0.1/lib/linux" ]
+    libs = [ ":libclang_rt.builtins-i386.a" ]
+  } else if (current_cpu == "arm") {
+    lib_dirs = [ "${sysroot}/usr/lib/llvm-13/lib/clang/13.0.1/lib/linux" ]
+    libs = [ ":libclang_rt.builtins-armhf.a" ]
+  }
+
   include_dirs = [
     "${tint_root_dir}/",
     "${tint_root_dir}/include/",
--- a/v8/BUILD.gn
+++ b/v8/BUILD.gn
@@ -14,6 +14,7 @@ import("//build_overrides/build.gni")
 
 import("gni/snapshot_toolchain.gni")
 import("gni/v8.gni")
+import("//build/config/sysroot.gni")
 
 # Specifies if the target build is a simulator build. Comparing target cpu
 # with v8 target cpu to not affect simulator builds for making cross-compile
@@ -700,6 +701,15 @@ config("internal_config_base") {
     "$target_gen_dir",
     "$target_gen_dir/include",
   ]
+
+  # work around a bug with clang not being able to find the __mulodi4 symbol on 32-bit archs
+  if (v8_current_cpu == "x86") {
+    lib_dirs = [ "${sysroot}/usr/lib/llvm-13/lib/clang/13.0.1/lib/linux" ]
+    libs = [ ":libclang_rt.builtins-i386.a" ]
+  } else if (v8_current_cpu == "arm") {
+    lib_dirs = [ "${sysroot}/usr/lib/llvm-13/lib/clang/13.0.1/lib/linux" ]
+    libs = [ ":libclang_rt.builtins-armhf.a" ]
+  }
 }
 
 config("internal_config") {
